src/dmd/dcache.d

```diff
@@ -0,0 +1,376 @@
+            owner = cast(shared(pid_t)*)mapping.ptr;
+            nextAlloc = cast(uint*)(mapping.ptr + CACHE_LINE);
+            toc = cast(TocEntry[])mapping[META_SIZE .. META_SIZE + TOC_SIZE];
+            data = cast(Data*)(mapping.ptr + META_SIZE + TOC_SIZE);
+            if (toc.length & (toc.length-1)) abort();
+            ourPid = getpid();
+            // Lock before modification of shared data
+            {
+                lock();
+                scope(exit) unlock();
+                // If first entry is all zeros - new mapping
+                if (data.occupied == 0 && data.keySize == 0)
+                {
+                    data.keySize = DATA_SIZE - Data.sizeof;
+                    foreach (ref t; toc) t.index = uint.max;
+                }
+            }
+        }
+
+        /// Clean the cache
+        void clear()
+        {
+            lock();
+            scope(exit) unlock();
+            (cast(ubyte[])mapping)[]  = 0;
+            data.keySize = DATA_SIZE - Data.sizeof;
+            foreach (ref t; toc) t.index = uint.max;
+        }
+
+        static struct TocEntry
+        {
+            uint hash32;
+            uint index; // == uint.max if empty
+            bool empty(){ return index == uint.max; }
+        }
+        static assert(TocEntry.sizeof == 8);
+
+        static struct Data
+        {
+            uint occupied; // 0 - free block, 1 - occupied
+            uint keySize;
+            uint valueSize;
+            // ubyte[keySize] - key
+            // ubyte[valueSize] - data
+            ubyte[] key()
+            {
+                assert(occupied);
+                assert(keySize > 0);
+                ubyte* k = cast(ubyte*)&this + Data.sizeof;
+                return k[0..keySize];
+            }
+
+            ubyte[] value()
+            {
+                assert(occupied);
+                assert(valueSize > 0);
+                ubyte* v = key.ptr + keySize;
+                return v[0..valueSize];
+            }
+
+            uint fullSize(){ return cast(uint)(Data.sizeof + keySize + valueSize); }
+        }
+        static assert(Data.sizeof == 12);
+
+        const(ubyte)[] get(const(ubyte)[] key)
+        {
+            uint h = first32bits(key);
+            size_t offset = h & (toc.length-1);
+            lock();
+            scope(exit) unlock();
+            for (;;)
+            {
+                if (toc[offset].empty) return null;
+                if (toc[offset].hash32 == h)
+                {
+                    auto p = blockAt(toc[offset].index);
+                    if (p.occupied && p.key == key)
+                    {
+                        return p.value;
+                    }
+                }
+                offset = (offset + 1) & (toc.length-1);
+            }
+        }
+
+        bool put(const(ubyte)[] key, const(ubyte)[] value)
+        {
+            uint h = first32bits(key);
+            size_t offset = h & (toc.length-1);
```

Comment:
I assume this means `toc.length` must be a power of 2. I don't see where this is documented.